sdk/storage/Azure.Storage.Blobs.Cryptography/src/ClientSideDecryptionPolicy.cs

```diff
@@ -0,0 +1,412 @@
+                    ciphertext.Read(IV, 0, IV.Length);
+                    read = IV.Length;
+                }
+
+                plaintext = WrapStream(
+                    ciphertext,
+                    GetContentEncryptionKey(encryptionData.Value).ToArray(),
+                    encryptionData.Value,
+                    IV,
+                    noPadding);
+            }
+            else
+            {
+                plaintext = ciphertext;
+            }
+
+            int gap;
+            if ((gap = (int)(encryptedBlobRange.OriginalRange.Offset - encryptedBlobRange.AdjustedRange.Offset) - read) > 0)
+            {
+                plaintext.Read(new byte[gap], 0, gap);
+            }
+
+            return new LengthLimitingStream(plaintext, encryptedBlobRange.OriginalRange.Count);
+        }
+
+        private async Task<Stream> DecryptBlobAsync(Stream ciphertext, Metadata metadata,
+            EncryptedBlobRange encryptedBlobRange, bool noPadding)
+        {
+            AssertKeyAccessPresent();
+            var encryptionData = GetAndValidateEncryptionData(metadata);
+
+            Stream plaintext;
+            int read = 0;
+            if (encryptionData.HasValue)
+            {
+                byte[] IV;
+                if (encryptedBlobRange.AdjustedRange.Offset == 0)
+                {
+                    IV = encryptionData.Value.ContentEncryptionIV;
+                }
+                else
+                {
+                    IV = new byte[EncryptionConstants.ENCRYPTION_BLOCK_SIZE];
+                    await ciphertext.ReadAsync(IV, 0, IV.Length).ConfigureAwait(false);
+                    read = IV.Length;
+                }
+
+                plaintext = WrapStream(
+                    ciphertext,
+                    (await GetContentEncryptionKeyAsync(encryptionData.Value).ConfigureAwait(false)).ToArray(),
+                    encryptionData.Value,
+                    IV,
+                    noPadding);
+            }
+            else
+            {
+                plaintext = ciphertext;
+            }
+
+            // still need to readjust ranges even if we didn't decrypt anything, so keep this out of branch
+            int gap;
+            if ((gap = (int)(encryptedBlobRange.OriginalRange.Offset - encryptedBlobRange.AdjustedRange.Offset) - read) > 0)
+            {
+                await plaintext.ReadAsync(new byte[gap], 0, gap).ConfigureAwait(false);
+            }
+
+            return new LengthLimitingStream(plaintext, encryptedBlobRange.OriginalRange.Count);
+        }
+
+        #endregion
+
+        private static Metadata ExtractMetadata(ResponseHeaders headers)
+        {
+            Metadata metadata = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase);
+            foreach (HttpHeader header in headers)
+            {
+                if (header.Name.StartsWith(Constants.HeaderNames.MetadataPrefix, StringComparison.InvariantCulture))
```

Comment:
Do you want `OrdinalIgnoreCase` as well here since this is also case insensitive?

Reply:
Interesting. The service documentation doesn't actually explain how the x-ms-meta- prefix interacts with the rest of the identifier being "case insensitive but preserving." We'll add it just in case, but I don't know if that's possible or not.